src/signature.c

```diff
@@ -54,111 +54,289 @@
 
 bool signing_enabled = false;
 
-
-static X509_STORE *create_store(const char *, const char *, const char *);
-
-static char *VERIF_FAIL = "Signature verification failed";
-static char *XSTORE_FAIL = "XSTORE creation failed";
+static bool validate_signature(FILE *, FILE *);
+static bool validate_certificate(void);
+static int verify_callback(int, X509_STORE_CTX *);
+static bool get_pubkey(const char *);
+static bool get_certificates(void);
 
 static bool initialized = false;
+static EVP_PKEY *pkey;
+static X509 *cert;
+static char *chain;
+static char *crl;
+static char *certificate;
+static char *ca_dirname;
 
-static X509_STORE *x509_store = NULL;
 
 bool signature_initialize(const char *ca_cert_filename)
 {
 	if (initialized) {
 		return true;
 	}
+
 	OpenSSL_add_all_algorithms();
 	ERR_load_crypto_strings();
-	x509_store = create_store(ca_cert_filename, NULL, NULL);
-	if (x509_store == NULL) {
+
+	if (!get_certificates()) {
+		return false;
+	}
+
+	if (!validate_certificate()) {
 		ERR_free_strings(); // undoes ERR_load_crypto_strings
 		EVP_cleanup();      // undoes OpenSSL_add_all_algorithms
 		return false || (IMPL == IMPL_FORGIVE);
 	}
+
+	if (!get_pubkey(ca_cert_filename)) {
+		return false || (IMPL == IMPL_FORGIVE);
+	}
+
 	initialized = true;
 	return true;
 }
 
 void signature_terminate(void)
 {
 	if (initialized) {
-		X509_STORE_free(x509_store); // undocumented...
 		ERR_free_strings();	  // undoes ERR_load_crypto_strings
-		EVP_cleanup();		     // undoes OpenSSL_add_all_algorithms
+		EVP_cleanup();		  // undoes OpenSSL_add_all_algorithms
+		EVP_PKEY_free(pkey);
+		X509_free(cert);
 		initialized = false;
 	}
 }
 
+bool get_certificates(void)
+{
+	/* TODO - get paths and certificate names from Linux environment */
+
```

Comment:
I would like the certs to live in a directory under /usr/share, perhaps in the same directory as the SSL certs. Though that directory name should not be Clear Linux specific, so we can rename it to /usr/share/swupd/certs, or similar.